docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,363 @@
+It is designed to be easy to implement in a language with JSON support.
+
+Protocol messages are separated by empty lines (double newline).
+The headers (request and response) are not JSON, but a sequence of space-separated values.
+All messages sent by cf-agent and the promise module are single line JSON-data, except:
+
+* Headers (both from cf-agent and promise module) are not JSON.
+* JSON responses sent from promise module may optionally be preceeded by log messages, as explained below.
+
+##### Example requests in JSON based protocol
+
+```
+CFEngine 3.16.0 v1
+
+{"operation": "validate_promise", "log_level": "info", "promise_type": "git", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}}
+
+{"operation": "evaluate_promise", "log_level": "info", "promise_type": "git", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}}
+
+{"operation": "terminate", "log_level": "info"}
+```
+
+##### Example response in JSON based protocol
+
+```
+git_promises 0.0.1 v1 json_based
+
+{"operation": "validate_promise", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}, "result": "valid"}
+
+log_info=Cloning 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'...
+log_info=Successfully cloned 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'
+{"operation": "evaluate_promise", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}, "result": "kept"}
+
+{"operation": "terminate", "result": "success"}
+```
+
+In the example above, log messages are sent to the agent on separate lines.
+The syntax is the same as for the line based protocol, described below.
+This is done so the agent can print the messages while the promise is evaluating.
+(In the example, cloning the repo takes a few seconds, and it is nice to get some feedback).
+
+Log messages formatted like this must be before the JSON message, and it is optional.
+You can also include log messages in the JSON data:
+
+```
+{
+  "operation": "evaluate_promise",
+  "promiser": "/opt/cfengine/masterfiles",
+  "attributes": {
+    "repo": "https://github.com/cfengine/masterfiles"}
+  },
+  "log": [
+    {
+      "level": "info",
+      "message": "Cloning 'https://github.com/cfengine/masterfiles' -> '/opt/cfengine/masterfiles'..."
+    }
+  ],
+  "result": "kept"
+}
+```
+
+(Formatted here for readability, JSON data must still be on a single line in protocol).
+
+#### Line based protocol
+
+The line based protocol supports a limited subset of features of CFEngine promises, which should be enough for most use cases.
+In particular, data passed to the module (through promise attributes) must be strings, without newlines.
+`slist`, `data` containers, strings with newlines are not supported.
```

Comment:
add the "and" in the list? "... containers, and strings ..." assuming you'd like to use the oxford comma. :)

Reply:
Thanks, fixed.